packages/core/src/types.ts

```diff
@@ -324,7 +323,25 @@ export function getElementDefinition(typeName: string, propertyName: string): In
   if (!typeSchema) {
     return undefined;
   }
-  return typeSchema.elements[propertyName] ?? typeSchema.elements[propertyName + '[x]'];
+
+  // Always try to match the exact property name first
+  const simpleMatch = typeSchema.elements[propertyName] ?? typeSchema.elements[propertyName + '[x]'];
+  if (simpleMatch) {
+    return simpleMatch;
+  }
+
+  // The propertyName can be a "choice of type" property, such as "value[x]", but in resolved form "valueString".
+  // So we need to iterate through all the elements and find the one that matches.
+  for (const [elementName, element] of Object.entries(typeSchema.elements)) {
+    for (const type of element.type) {
+      if (elementName.replace('[x]', capitalize(type.code)) === propertyName) {
+        return element;
+      }
+    }
+  }
```

Comment:
This is kinda inefficient for a function that might be called in some hot loops; would it be possible to try and find the correct definition directly by doing something like this?

```ts
const fieldParts = propertyName.split(/(?=[A-Z])/g);
for (let i = 0; i < fieldParts.length; i++) {
  const testProperty = fieldParts.slice(0, i+1).join('') + '[x]';
  const element = typeSchema.elements[testProperty];
  if (element) {
    return element;
  }
}
```

Reply:
Neat, that is definitely faster.  That was counterintuitive to me due to the regex split. 

Old (not supporting `valueBoolean`): 1028ms
As implemented in PR: 1549ms
As implemented in suggestion: 1122ms

Strategy: For each, run 5 times, take the best time.

Test harness: 

```ts
import { readJson } from '@medplum/definitions';
import { Bundle } from '@medplum/fhirtypes';
import { indexStructureDefinitionBundle } from '../typeschema/types';
import { getTypedPropertyValue } from './utils';

indexStructureDefinitionBundle(readJson('fhir/r4/profiles-types.json') as Bundle);
indexStructureDefinitionBundle(readJson('fhir/r4/profiles-resources.json') as Bundle);

const startTime = Date.now();
let count = 0;

for (let i = 0; i < 100000; i++) {
  for (const propName of ['valueBoolean', 'value[x]', 'asdf']) {
    if (getTypedPropertyValue({ type: 'Extension', value: { [propName]: true } }, propName)) {
      count++;
    }
  }
}

const endTime = Date.now();
const duration = endTime - startTime;
console.log('Elapsed time: ' + duration + 'ms');
console.log('Count: ' + count);
```

Reply:
Nerd sniped by optimization problem: 1074ms

```ts
  for (let i = 0; i < propertyName.length; i++) {
    const c = propertyName[i];
    if (c >= 'A' && c <= 'Z') {
      const testProperty = propertyName.slice(0, i) + '[x]';
      const element = typeSchema.elements[testProperty];
      if (element) {
        return element;
      }
    }
  }
```

Reply:
Love it — that's a full 30% faster than the O(N^2) solution even for a small input! 🔥